nvdaHelper/remote/apiHook.h

```diff
@@ -16,38 +16,33 @@ This license can be found at:
 #define _APIHOOK_H
 
 /**
- * Initializes API hooking subsystem.
+ * Initializes an API hooking transaction.
  * @return success flag
  */ 
-bool apiHook_initialize();
+bool apiHook_beginTransaction();
 
 /**
  * Requests that the given function from the given module should be hooked with the given hook procedure. 
  * Warning, this function has no safety checks, you should usually use the apiHook_hookFunction_safe macro
- * @param moduleName the name of the module the function you wish to hook is located in.
- * @param functionName the name of the function you wish to hook.
- * @param newHookProc the function you wish  to be called instead of the original one.
- * @return the address of the original function. You could use this to call the origianl function from with in your replacement.
+ * @param realFunction the name of the function you wish to hook.
+ * @param fakeFunction the function you wish  to be called instead of the original one.
+ * @param targetPointerRef Pointer to the target pointer to which the detour will be attached.
  */ 
-void* apiHook_hookFunction(const char* moduleName, const char* functionName, void* newHookProc);
-
- /**
- * a helper template used internally by apiHook_hookFunction_safe
- */
-template<typename funcType> funcType _apiHook_hookFunction_tpl(const char* moduleName, const char* functionName, funcType funcSyg, funcType fakeFunction) { return (funcType)apiHook_hookFunction(moduleName,functionName,(void*)fakeFunction); }
+bool apiHook_hookFunction(void* realFunction, void* fakeFunction, void** targetPointerRef);
 
 /**
  * Safely hooks a given function from a given module with a given fake function.
- * @param moduleName a string containing the name of the module the function lives in
- * @param the name/symbol of the function that should be hooked (i.e. the symbole from its header file, not a string) -- the the symbol's type is used for safety, the the function name string looked up in the module is made from this symbol name.
- * @param fakeFunction the replacement function.
- */ 
-#define apiHook_hookFunction_safe(moduleName,realFunction,fakeFunction) _apiHook_hookFunction_tpl(moduleName,#realFunction,realFunction,fakeFunction)
- 
+ * @param realFunction the name of the function you wish to hook.
+ * @param fakeFunction the function you wish  to be called instead of the original one.
+ * @param targetPointerRef Pointer to the target pointer to which the detour will be attached.

```

Comment:
I'm not sure this is accurate, this just seems to be a way to maintain access to the original function (by saving its address). Look at the implementation in `nvdaHelper/remote/apiHook.cpp:52`, it overrides this with the `realFunction`. 

The same pattern used by Detours: https://github.com/microsoft/Detours/wiki/Using-Detours

Given this, we don't really need the realFunction argument, though it is helpful for type checking, making it harder to mess up.